src/main/kotlin/uk/gov/dwp/dataworks/kafkatohbase/reconciliation/repositories/impl/MetadataStoreRepositoryImpl.kt

```diff
@@ -0,0 +1,157 @@
+                                  private val table: String) : MetadataStoreRepository {
+
+    override fun groupedUnreconciledRecords(): Map<String, List<UnreconciledRecord>> =
+        unreconciledRecords().groupBy { it.topicName }
+
+    override fun reconcileRecord(topicName: String, hbaseId: String, version: Long) {
+        val rowsInserted = markRecordAsReconciled(topicName, hbaseId, version)
+        logger.info("Recorded processing attempt", "topic_name" to topicName, "rows_inserted" to "$rowsInserted")
+    }
+
+    override fun reconcileRecords(unreconciled: List<UnreconciledRecord>) {
+        if (unreconciled.isNotEmpty()) {
+            logger.info("Reconciling records", "record_count" to "${unreconciled.size}")
+            with (reconcileRecordStatement) {
+                try {
+                    unreconciled.forEach {
+                        setInt(1, it.id)
+                        addBatch()
+                    }
+                    executeBatch()
+                    commit()
+                } catch (e: SQLException) {
+                    logger.error("Failed to update batch", e, "error" to "e.message", "error_code" to "${e.errorCode}")
+                    rollback()
+                }
+            }
+            logger.info("Reconciled records", "record_count" to "${unreconciled.size}")
+        }
+    }
+
+    override fun fetchUnreconciledRecords(): List<Map<String, Any>> {
+        logger.debug("Fetching unreconciled records from metadata store")
+        val unreconciledRecords = getUnreconciledRecordsQuery()
+        return mapResultSet(unreconciledRecords)
+    }
+
+    private fun unreconciledRecords(): MutableList<UnreconciledRecord> {
+        val list = mutableListOf<UnreconciledRecord>()
+        unreconciledRecordsStatement.executeQuery().use {
+            while (it.next()) {
+                list.add(UnreconciledRecord(it.getInt("id"),
+                                            it.getString("topic_name"),
+                                            it.getString("hbase_id"),
+                                            it.getTimestamp("hbase_timestamp").time))
+            }
+        }
+        return list
+    }
+
+    private fun getUnreconciledRecordsQuery(): ResultSet? {
+        val statement = connection.createStatement()
+        return statement.executeQuery(
+            """
+                SELECT * FROM $table
+                WHERE write_timestamp > CURRENT_DATE - INTERVAL 14 DAY AND 
+                reconciled_result = false
+                LIMIT 100000 
+            """.trimIndent()
+        )
+    }
+
+
+    private val unreconciledRecordsStatement: PreparedStatement by lazy {
+        connection.prepareStatement("""
+                SELECT id, hbase_id, hbase_timestamp, topic_name 
+                FROM $table
+                WHERE write_timestamp > CURRENT_DATE - INTERVAL 14 DAY 
+                AND reconciled_result = false
+                LIMIT 100000
```

Comment:
Don't need to change now but this will merge into $queryLimit

Reply:
Gotcha